lib/pbench/cli/agent/commands/generate_token.py

```diff
@@ -0,0 +1,71 @@
+"""pbench-generate-token"""
+
+import click
+import json
+import requests
+
+from pbench.agent.base import BaseCommand
+from pbench.cli.agent import pass_cli_context
+from pbench.cli.agent.options import common_options
+
+
+class GenerateToken(BaseCommand):
+    def __init__(self, context):
+        super().__init__(context)
+
+    def execute(self):
+        """Generate a token used to access the pbench server RESTful API"""
+
+        # FIXME:  Should we be logging anything here?
+        #
+        #  self.logger.warn(
+        #     "%s deprecated, will be removed in future release in favor of pbench-clear-results",
+        #     self.name,
+        # )
+
+        server = self.config.agent.parser.get("results", "server_rest_url")
+        headers = {"Content-Type": "application/json", "Authorization": None}
+        payload = {
+            "username": self.context.username,
+            "password": self.context.password,
+            "token_duration": self.context.token_duration
```

Comment:
`"token_duration"` isn't a parameter of Nikhil's login code. Did you change that? I don't think we want users to specify their own token duration, even if we're bounding it with our own limitations. I think we probably do want to add a "role" request in some form, and here we'd be asking for "PUT-only", and the role might potentially imply a duration (with more restricted roles potentially having a longer duration).

Reply:
I believe that we are going to want to allow the user of this tool to be able to specify the token duration.  Obviously, it will be up to the server to decide what to do about the request (e.g., impose a cap, ignore it altogether, etc.), so I figured it wouldn't hurt (much) to go ahead an implement it in the client end.  (That is, I fully expect that specifying this value will have no effect initially.)  Given that the server doesn't actually _require_ a token, yet, we're in a realm of "total flexibility"; as we converge upon the final implementation, I'm hoping we can fold flexible duration into the mix.

AWS S3 offers capability analogous to this, allowing the client to specify how long a pre-signed URL is good for.  I can envision use cases where the client needs only minutes, and I can envision other cases where the client will want the token to be good for as long as the server will tolerate -- I don't see any reason why we cannot and should not allow the user to choose.

Your point about allowing the request to specify a role is an excellent one!  Once we have a better handle on what roles look like, we can extend this tool.  (We had originally envisioned this tool's request as receiving a long-duration `PUT`-only token, but we currently have no way to effect that, without adding to the server API, and I would rather have that extension provide support for roles rather than custom support for move-results.)  For now, I think that we should have this tool execute a standard login and receive a standard token with a standard duration, with the server simply ignoring the request for a different token duration.

Reply:
yeah `token_duration` is not a part of user defined parameter only username and password

Reply:
I'm not sure I can ever see someone realistically saying "I know I won't want to upload data after more than 10 minutes so let's limit the PUT token I'm using for my cluster deployment tests to 10 minutes". I can't imagine anyone specifying less than the maximum, really, because "things happen" and going back to the well for a second bucket in the middle of putting out the fire is a pain nobody really wants.

But at the very least, I'd prefer to see a TODO comment here explaining that this is currently meaningless but is added as a stake in the ground for possible later experimentation. (Though I think it's just as easy to remove it now and put it or something else in later when we figure out what that is. An issue to record current thoughts on that future seems like a more effective substitute for non-operative code.)

Reply:
If the server had a broader exposure than the RD intranet, I suspect a user wouldn't want their tokens to last any longer than needed; on the flip side, I can easily imagine that different users would have different requirements for how long they need their tokens to live.  So, I "one size" solution is unlikely to fit "all".

I can certainly add a TODO and open an issue.  While it's not hard to remove and re-add this capability, I'm inclined to leave it against the day when the server starts supporting it.  (I really don't think that the tokens intended to support human use of the dashboard are going to work very well for scripted transfer of benchmark results.)

Reply:
`TODO` added; issue #2139 opened.